### PR TITLE
Fix earlier/later missing changeset update

### LINKED ISSUE #1059: Undo redo panic
Helix sometimes produces a panic after using undo redo.

minimal reproduction:

1. open helix
2. enter <kbd>i</kbd> <kbd>a</kbd> <kbd>Esc</kbd> <kbd>u</kbd> `:later` <kbd>o</kbd>

Interestingly, when using `:earlier` instead of <kbd>u</kbd> the panic does not occur.

---

<details>
<summary>backtrace</summary>

```
thread 'main' panicked at 'assertion failed: self.len_after == other.len', helix-core\src\transaction.rs:128:9
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:517
   1: core::panicking::panic_fmt
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:101
   2: core::panicking::panic
             at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:50
   3: helix_core::transaction::ChangeSet::compose
   4: helix_view::document::Document::apply
   5: helix_term::commands::append_to_line
   6: helix_term::ui::editor::EditorView::render_statusline
   7: <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::handle_event
   8: helix_term::application::Application::handle_terminal_events
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: tokio::park::thread::CachedParkThread::block_on
  11: tokio::runtime::thread_pool::ThreadPool::block_on
  12: tokio::runtime::Runtime::block_on
  13: hx::setup_logging
```
</details>

### PR DESCRIPTION
**Pull request by pickfire**
Fix #1059

cc @CptPotato

### CODE REVIEW

File: helix-view/src/document.rs [hunk 1]
```diff
@@ -751,19 +751,35 @@ impl Document {
...
+        let mut success = false;
         for txn in txns {
-            self.apply_impl(&txn, view_id);
+            if self.apply_impl(&txn, view_id) {
+                success = true;
+            }
+        }
+        if success {
+            // reset changeset to fix len
+            self.changes = ChangeSet::new(self.text());
```
Comment (archseer) [inline helix-view/src/document.rs:780]: Good catch!

**Comment by CptPotato** [comment]
I'm not able to reproduce the panic with this fix anymore, thanks 👍 

On a side note, I think I misunderstood `:earlier` and `:later`. I initially thought they would be identical to undo and redo but they actually work differently. Is this the intended behaviour?

**Comment by archseer** [comment]
@CptPotato I personally don't use these (though I should try to), but they hook into the undo tree so you're able to go to a version of the document at a specific time (could be on a different undo branch): https://vimtricks.com/p/vimtrick-time-travel-in-vim/
https://github.com/helix-editor/helix/blob/d131a9dd0efc5ff271f8b78cd65a8dc30c193af4/helix-core/src/history.rs#L28-L32

**Comment by CptPotato** [comment]
> @CptPotato I personally don't use these (though I should try to), but they hook into the undo tree so you're able to go to a version of the document at a specific time (could be on a different undo branch): https://vimtricks.com/p/vimtrick-time-travel-in-vim/

That's really cool, thank you for elaborating.

**Pull request merged**